sources/ElkArte/Graphics/Gd2.php

```diff
@@ -0,0 +1,494 @@
+<?php
+
+/**
+ * This file deals with low-level graphics operations performed on images,
+ * specially as needed for the GD library
+ *
+ * @package   ElkArte Forum
+ * @copyright ElkArte Forum contributors
+ * @license   BSD http://opensource.org/licenses/BSD-3-Clause (see accompanying LICENSE.txt file)
+ *
+ * @version 2.0 dev
+ *
+ */
+
+namespace ElkArte\Graphics;
+
+/**
+ * Class Gd2
+ *
+ * @package ElkArte\Graphics
+ */
+class Gd2 extends AbstractManipulator
+{
+	/** @var resource */
+	protected $_image;
+
+	/**
+	 * Gd2 constructor.
+	 *
+	 * @param string $image
+	 */
+	public function __construct($image)
+	{
+		$this->setSource($image);
+
+		try
+		{
+			$this->memoryCheck();
+		}
+		catch (\Exception $e)
+		{
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Set the filename to the class
+	 *
+	 * @param $source
+	 */
+	public function setSource($source)
+	{
+		$this->_fileName = $source;
+	}
+
+	/**
+	 * Used to determine if the GD2 library is present.
+	 *
+	 * @return bool Whether or not GD is available.
+	 */
+	public static function canUse()
+	{
+		// Check to see if GD is installed and what version.
+		if ((get_extension_funcs('gd')) === false)
+		{
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Cleanup when done
+	 */
+	public function __destruct()
+	{
+		if (is_resource($this->_image))
+		{
+			imagedestroy($this->_image);
+		}
+	}
+
+	/**
+	 * Loads a image file into the image engine for processing
+	 *
+	 * @return bool|mixed
+	 */
+	public function createImageFromFile()
+	{
+		$this->getSize();
+
+		if (isset(Image::DEFAULT_FORMATS[$this->sizes[2]]))
+		{
+			try
+			{
+				$imagecreatefrom = 'imagecreatefrom' . Image::DEFAULT_FORMATS[$this->sizes[2]];
+				$image = $imagecreatefrom($this->_fileName);
+			}
+			catch (\Exception $e)
+			{
+				return false;
+			}
+		}
+		else
+		{
+			return false;
+		}
+
+		$this->_setImage($image);
+
+		return true;
+	}
+
+	/**
+	 * Sets the internal GD image resource.
+	 *
+	 * @param resource $image
+	 */
+	protected function _setImage($image)
+	{
+		$this->_image = $image;
+
+		// Get the image size via GD functions
+		$this->_width = $this->sizes[0] = imagesx($image);
+		$this->_height = $this->sizes[1] = imagesy($image);
+	}
+
+	/**
+	 * Loads an image from a web address into the image engine for processing
+	 *
+	 * @return bool|mixed
+	 */
+	public function createImageFromWeb()
+	{
+		require_once(SUBSDIR . '/Package.subs.php');
+		$this->_image = fetch_web_data($this->_fileName);
```

Comment:
$this->_image should be $image I think

Reply:
Its a cheat  .... $this->getSize expects its data to be available in $this->_image, so although at that point its not an image but a string,  I did not want to pass the value.